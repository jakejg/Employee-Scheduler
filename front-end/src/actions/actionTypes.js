export const LOAD_JOBS = "LOAD_JOBS";
export const ADD_JOBS = "ADD_JOB";