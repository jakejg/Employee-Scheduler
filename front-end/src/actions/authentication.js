import axios from 'axios';
import { 
    ADD_OR_REMOVE_TOKEN
 } from './actionTypes';
import {addCompanyToAPI} from '../helpers/JobApi'
import {BASE_URL} from '../config';


export const register = (data) => {
    
    return async (dispatch) => {
        try {
            // designate admin user
            data.is_admin = true
            // create admin user
       
            const res = await axios.post(`${BASE_URL}/auth/register`, data);
       
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const login = (data) => {
    return async (dispatch) => {
        try {
            let res = await axios.post(`${BASE_URL}/auth/login`, data);
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            // console.log(e.response.data.message)
            return e.response.data.message
        }
    }
}

export const addOrRemoveToken = (token) => {
    return {type: ADD_OR_REMOVE_TOKEN, token}
}
