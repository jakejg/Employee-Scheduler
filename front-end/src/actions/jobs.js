import { 
    LOAD_JOBS,
    EDIT_JOB, 
    ADD_JOB,
    EDIT_JOB_STAFF,
    DELETE_JOB
 } from './actionTypes';
import {JobAPI} from '../helpers/JobApi';

export const loadJobsFromAPI = (comp_id) => {
    return async (dispatch) => {
        try {
            const jobData = await JobAPI.loadJobs(comp_id)
            dispatch(loadJobs(jobData));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const loadJobs = (jobs) => {
    return {type: LOAD_JOBS, jobs}
}

export const addJobOnAPI = (jobToAdd) => {

    return async (dispatch) => {
        try {
            let job = await JobAPI.addJob(jobToAdd)
            dispatch(addJob(job.id, job))
            return {message: "Job created Successfully", severity: "success"}
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const addJob = (id, job) => {
    return {type: ADD_JOB, id, job}
}


export const getJobFromAPI = (ID) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.getJob(ID)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJobOnAPI = (id, changedProps) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.editJob(id, changedProps)
            console.log(job)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJob = (id, job) => {
    return {type: EDIT_JOB, id, job}
}

export const addStaffToJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.addStaffToJob(jobId, staffId)
            console.log(data)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const removeStaffFromJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.removeStaffFromJob(jobId, staffId)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const editJobStaff = (id, staff, staff_filled) => {
    return {type: EDIT_JOB_STAFF, id, staff, staff_filled}
}

export const deleteJobOnAPI = (id) => {
    return async (dispatch) => {
        try {
            let msg = await JobAPI.deleteJob(id)
            dispatch(deleteJob(id))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const deleteJob = (id) => {
    return {type: DELETE_JOB, id}
}