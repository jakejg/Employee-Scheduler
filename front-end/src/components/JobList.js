import React from 'react';
import { useSelector } from 'react-redux';
import {Link} from 'react-router-dom';
import {List, ListItem, ListItemText, Box, Typography } from '@material-ui/core';
import moment from 'moment';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(() => ({
    link: {
      textDecoration: 'none',
      color: 'black',
      '&:hover': {
        color: 'grey'
      }
    }
}))


const JobList = () => {
    const jobs = useSelector(state => state.jobs)
    const classes = useStyles();
    
    let inProgress = {};
    let scheduled = {};
    for (let key in jobs){
        if (jobs[key].start_time > moment()){
            scheduled[key] = jobs[key]
        }
        else if (jobs[key].end_time > moment()){
            inProgress[key] = jobs[key]
        }
        
    }
    
    return (
        <div>
            <Typography variant='h4' align='center'>Jobs in Progress</Typography>
            <List>
                {Object.keys(inProgress).map(id => <ListItem key={id}>
                                                        <ListItemText align='center'>
                                                            <Link className={classes.link} to={`job/${id}`}>{inProgress[id].title}</Link>
                                                        </ListItemText>
                                                    </ListItem>)}
            </List>
            <Typography variant='h4' align='center'>Jobs Scheduled</Typography>
            <List>
                {Object.keys(scheduled).map(id => <ListItem key={id}>
                                                        <ListItemText align='center'>
                                                        <Link className={classes.link} to={`job/${id}`}>{scheduled[id].title}</Link>
                                                        </ListItemText>
                                                    </ListItem>)}
            </List>

        </div>
    );
}

export default JobList;
