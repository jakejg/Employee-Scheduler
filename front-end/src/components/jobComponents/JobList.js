import React from 'react';
import { useSelector} from 'react-redux';
import {List, ListItem, Box, Typography, makeStyles, Chip, Tooltip} from '@material-ui/core';
import DoneOutlineRoundedIcon from '@material-ui/icons/DoneOutlineRounded';
import moment from 'moment';
import Loading from '../Loading';
import statusToColor from '../../helpers/jobColorObj';

const useStyles = makeStyles(() => ({
    green: {
        color: 'green', 
        marginLeft: '5px', 
        marginTop: '5px'
    },
    yellow: {
        color: '#f0ad4e', 
        marginLeft: '5px', 
        marginTop: '5px'
    },
    red: {
        color: 'red', 
        marginLeft: '5px', 
        marginTop: '5px'
    },
    text: {
        display: 'flex',
        justifyContent: 'start',
        width: '125px'
    },
    box: {
        display: 'flex',
        justifyContent: 'center',
    }
}))


const JobList = () => {
    const jobs = useSelector(state => state.jobs)
    const classes = useStyles();
    const loading = !Object.keys(jobs).length
    const statusColor = statusToColor();

    if (loading) return <Loading />

    let inProgress = {};
    let scheduled = {};
    for (let key in jobs){
        if (jobs[key].start_time > moment()){
            scheduled[key] = jobs[key]
        }
        else if (jobs[key].end_time > moment()){
            inProgress[key] = jobs[key]
        }
    }
    const getText = (type,id) => {
        let toolTipText;
        if (type[id].status === 'under'){
            toolTipText = 'Need Staff';
        }
        else if (type[id].status === 'filled'){
            toolTipText = 'Staff Filled';
        }
        else {
            toolTipText = 'Over Staffed';
        }
        return toolTipText;
    }

    const getStaffNames = (staffList) => {
        if (!staffList.length) return ""

        return (<List disablePadding={true}>
                    {staffList.map(staff => <ListItem key={staff.id} dense={true}>
                                                {`${staff.first_name} ${staff.last_name}`}
                                            </ListItem>)}
                </List>)
    }
    
    return (
        <div>
            <Typography variant='h4' align='center'>Jobs in Progress</Typography>
            <List>
                {Object.keys(inProgress).map(id => <ListItem key={id} className={classes.box}>
                                                        <Box  className={classes.text}> 
                                                            <Tooltip title={getStaffNames(inProgress[id].staff)} placement="left">    
                                                                <Chip 
                                                                label={inProgress[id].title} 
                                                                component="a" href={`job/${id}`}
                                                                color="primary"
                                                                clickable />
                                                            </Tooltip>    
                                                            <Tooltip title={getText(inProgress, id)}>
                                                                <DoneOutlineRoundedIcon fontSize='small' className={classes[statusColor[inProgress[id].status]]}/>
                                                            </Tooltip>
                                                        </Box>
                                                    </ListItem>)}
            </List>
            <Typography variant='h4' align='center'>Jobs Scheduled</Typography>
            <List >
                {Object.keys(scheduled).map(id => <ListItem key={id} className={classes.box}>
                                                        <Box  className={classes.text}>
                                                            <Tooltip title={getStaffNames(scheduled[id].staff)} placement="left">
                                                                <Chip 
                                                                    label={scheduled[id].title} 
                                                                    component="a" href={`job/${id}`}
                                                                    color="primary"
                                                                    clickable
                                                                    />
                                                            </Tooltip>
                                                            <Tooltip title={getText(scheduled, id)}>
                                                                <DoneOutlineRoundedIcon fontSize='small' className={classes[statusColor[scheduled[id].status]]}/>
                                                            </Tooltip>
                                                        </Box> 
                                                    </ListItem>)}
            </List>

        </div>
    );
}

export default JobList;
