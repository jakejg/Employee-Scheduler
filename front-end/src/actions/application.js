import {CHANGE_DRAWER, CHANGE_NO_JOBS, CHANGE_NO_STAFF} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}

export const changeNoJobs = () => {
    return {type: CHANGE_NO_JOBS}
}

export const changeNoStaff = () => {
    return {type: CHANGE_NO_STAFF}
}