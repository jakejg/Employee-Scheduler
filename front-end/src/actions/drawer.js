import {CHANGE_DRAWER} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}