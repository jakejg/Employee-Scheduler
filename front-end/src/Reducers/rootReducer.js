import jobs from './jobsReducer';
import staff from './staffReducer';
import { combineReducers } from "redux";

export default combineReducers({jobs, staff})