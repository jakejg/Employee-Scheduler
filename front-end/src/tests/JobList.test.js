import React from 'react';
import { render } from '@testing-library/react';
import {Provider} from 'react-redux'
import JobList from '../components/JobList';
import {createStore, applyMiddleware} from "redux";
import rootReducer from "../reducers/rootReducer";
import { MemoryRouter } from "react-router";
import { loadJobs } from '../actions/jobs';
import thunk from 'redux-thunk'
import moment from 'moment';

const jobs = {
    1 : {
        start_time : moment('2020-10-04'),
        end_time : moment('2020-10-24'),
        group : 1,
        staff: [],
        status: 'under',
        title : '20 Day River'
    },
    2 : {
        start_time : moment('2020-10-15'),
        end_time : moment('2020-10-30'),
        group : 2,
        staff: [],
        status: 'under',
        title : '15 Day Mountain'
    }
}

const store = createStore( rootReducer, applyMiddleware(thunk));
store.dispatch(loadJobs(jobs)); 

it('renders without crashing', () => {

  render(
    <MemoryRouter>
      <Provider store={store}>
      <JobList />
      </Provider>
    </MemoryRouter>
)
});

it('matches snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <Provider store={store}>
                <JobList />
            </Provider>
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})


it('renders a list of jobs', () => {
    const { getByText } = render(
        <MemoryRouter>
            <Provider store={store}>
                <JobList />
            </Provider>
        </MemoryRouter>
    )
    expect(getByText('20 Day River')).toBeInTheDocument();
    expect(getByText('15 Day Mountain')).toBeInTheDocument();
})