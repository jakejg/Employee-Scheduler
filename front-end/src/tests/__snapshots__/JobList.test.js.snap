// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter"
    >
      Jobs in Progress
    </h4>
    <ul
      class="MuiList-root MuiList-padding"
    />
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter"
    >
      Jobs Scheduled
    </h4>
    <ul
      class="MuiList-root MuiList-padding"
    >
      <li
        class="MuiListItem-root makeStyles-box-12 MuiListItem-gutters"
      >
        <div
          class="MuiBox-root MuiBox-root-13 makeStyles-text-11"
        >
          <a
            class="MuiChip-root MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable"
            href="job/1"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiChip-label"
            >
              20 Day River
            </span>
          </a>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-yellow-9 MuiSvgIcon-fontSizeSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20.47 5.63c.39.39.39 1.01 0 1.4L9.13 18.37c-.39.39-1.01.39-1.4 0l-4.2-4.2a.9839.9839 0 010-1.4c.39-.39 1.01-.39 1.4 0l3.5 3.5L19.07 5.63c.39-.39 1.01-.39 1.4 0zm-2.11-2.12l-9.93 9.93-2.79-2.79c-.78-.78-2.05-.78-2.83 0l-1.4 1.4c-.78.78-.78 2.05 0 2.83l5.6 5.6c.78.78 2.05.78 2.83 0L22.59 7.74c.78-.78.78-2.05 0-2.83l-1.4-1.4c-.79-.78-2.05-.78-2.83 0z"
            />
          </svg>
        </div>
      </li>
      <li
        class="MuiListItem-root makeStyles-box-12 MuiListItem-gutters"
      >
        <div
          class="MuiBox-root MuiBox-root-14 makeStyles-text-11"
        >
          <a
            class="MuiChip-root MuiChip-colorPrimary MuiChip-clickableColorPrimary MuiChip-clickable"
            href="job/2"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiChip-label"
            >
              15 Day Mountain
            </span>
          </a>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-yellow-9 MuiSvgIcon-fontSizeSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20.47 5.63c.39.39.39 1.01 0 1.4L9.13 18.37c-.39.39-1.01.39-1.4 0l-4.2-4.2a.9839.9839 0 010-1.4c.39-.39 1.01-.39 1.4 0l3.5 3.5L19.07 5.63c.39-.39 1.01-.39 1.4 0zm-2.11-2.12l-9.93 9.93-2.79-2.79c-.78-.78-2.05-.78-2.83 0l-1.4 1.4c-.78.78-.78 2.05 0 2.83l5.6 5.6c.78.78 2.05.78 2.83 0L22.59 7.74c.78-.78.78-2.05 0-2.83l-1.4-1.4c-.79-.78-2.05-.78-2.83 0z"
            />
          </svg>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
